File: backend-go/internal/controller/messMenu.go
package controller

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"strconv"

	helpers "github.com/LambdaIITH/Dashboard/backend/internal/helpers"

	"github.com/gin-gonic/gin"
)

type menuWeekChangeRequest struct {
	Password string `json:"password"`
	Number   int    `json:"number"`
}

var admins = []string{"lambda@iith.ac.in", "ms22btech11010@iith.ac.in", "ma22btech11003@iith.ac.in", "cs22btech11017@iith.ac.in", "cs22btech11028@iith.ac.in"}
var allowedNumbers = []int{0, 1, 2, 3}

// GetMessMenu handles the GET request to retrieve the mess menu.
func GetMessMenu(c *gin.Context) {
	dir, err := os.Getwd()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
		return
	}

	// Read the current week from config.json
	configPath := filepath.Join(dir, "config.json")
	var config map[string]interface{}

	configFile, err := os.Open(configPath)
	if err != nil {
		if os.IsNotExist(err) {
			config = map[string]interface{}{"week": 0}

			file, err := os.Create(configPath)
			if err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create config.json"})
				return
			}
			defer file.Close()

			if err := json.NewEncoder(file).Encode(config); err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to write to config.json"})
				return
			}
		} else {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error: error accessing config.json"})
			return
		}
	} else {
		defer configFile.Close()

		// Read the existing configuration
		if err := json.NewDecoder(configFile).Decode(&config); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error: invalid config.json"})
			return
		}
	}

	// reading week number from config.json
	week, _ := config["week"].(float64)
	weekStr := strconv.FormatFloat(week, 'f', -1, 64)

	// Read the menu file for the current week
	menuPath := filepath.Join(dir, weekStr+".json")
	menuFile, err := os.Open(menuPath)
	if err != nil {
		fmt.Println(err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error: missing menu file"})
		return
	}
	defer menuFile.Close()

	var menu interface{}
	if err := json.NewDecoder(menuFile).Decode(&menu); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error: invalid menu file"})
		return
	}

	c.JSON(http.StatusOK, menu)
}

// PostMessMenu handles the POST request to update the mess week.
func PostMessMenu(c *gin.Context) {
	var req menuWeekChangeRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	// Check if the provided week number is valid
	isAllowed := false
	for _, num := range allowedNumbers {
		if num == req.Number {
			isAllowed = true
			break
		}
	}
	if !isAllowed {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid week number"})
		return
	}

	// Check if the user is an admin or has the correct password
	isAdmin := false
	token, err := c.Cookie("session")
	if err == nil {
		status, payload := helpers.VerifyIDToken(c, token) // verify token and get email
		if status {
			if email, ok := payload["email"].(string); ok {
				for _, admin := range admins {
					if email == admin {
						isAdmin = true
						break
					}
				}
			}
		}
	}

	// Verify the password if the user is not an admin
	password := os.Getenv("ADMIN_PASS")
	if !isAdmin && req.Password != password {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Unauthorized: incorrect password"})
		return
	}

	// Update the week in config.json
	dir, err := os.Getwd()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
		return
	}
	configPath := filepath.Join(dir, "config.json")
	config := map[string]int{"week": req.Number}

	// create new or overwrite existing file config.json
	file, err := os.Create(configPath)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update configuration file"})
		return
	}
	defer file.Close()

	// writing the updated week number to config.json
	if err := json.NewEncoder(file).Encode(config); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to write configuration data"})
		return
	}

	// return a JSON response
	c.JSON(http.StatusOK, gin.H{"message": "Week number updated successfully"})
}

// GetCurrentWeekNumber handles the GET request to retrieve the current week number.
func GetCurrentWeekNumber(c *gin.Context) {
	isAdmin := false

	// Validate the user's admin status
	token, err := c.Cookie("session")
	if err == nil {
		status, payload := helpers.VerifyIDToken(c, token)
		if status {
			if email, ok := payload["email"].(string); ok {
				for _, admin := range admins {
					if email == admin {
						isAdmin = true
						break
					}
				}
			}
		}
	}

	if !isAdmin {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Unauthorized"})
		return
	}

	dir, err := os.Getwd()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
		return
	}

	// Read the current week from config.json
	configPath := filepath.Join(dir, "config.json")
	configFile, err := os.Open(configPath)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error: missing config.json"})
		return
	}
	defer configFile.Close()

	// decode the JSON file to read week number
	var config map[string]interface{}
	if err := json.NewDecoder(configFile).Decode(&config); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error: invalid config.json"})
		return
	}

	week, _ := config["week"].(float64)

	// send week number in response
	c.JSON(http.StatusOK, gin.H{"week": int(week)})
}

File: backend/Routes/MessMenu/mess_menu.py
from fastapi import APIRouter, HTTPException, Depends, Request, BackgroundTasks
import json
import os
import datetime
import subprocess
from pydantic import BaseModel
from Routes.Auth.cookie import get_user_id
from Routes.User.user import get_user
from Routes.Auth.tokens import verify_token

router = APIRouter(prefix="/mess_menu", tags=["mess_menu"])
password = os.getenv("ADMIN_PASS")
allowed_numbers = [0,1,2,3]

admins = ["ms22btech11010@iith.ac.in", "lambda@iith.ac.in", "ma22btech11003@iith.ac.in", "cs22btech11017@iith.ac.in", "cs22btech11028@iith.ac.in"]

SHEETS_WEBHOOK_TOKEN = os.getenv("SHEETS_WEBHOOK_TOKEN")

# -------------------------------------------------------------------------------
# Timezone: IST (Asia/Kolkata)
try:
    IST = datetime.timezone(datetime.timedelta(hours=5, minutes=30))
except AttributeError:
    IST = datetime.timezone(datetime.timedelta(hours=5, minutes=30))

def now_ist() -> datetime.datetime:
    return datetime.datetime.now(IST)

def today_ist() -> datetime.date:
    return now_ist().date()

DAYS = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]

def get_today_name() -> str:
    """Get today's day name matching our DAYS array (Sunday=0) in IST."""
    py_wd = today_ist().weekday()  # Mon=0...Sun=6
    return DAYS[0 if py_wd == 6 else py_wd + 1]

def log_api(msg: str):
    """Log API events to menu_scraper.log for unified observability."""
    try:
        log_path = os.path.join(os.path.dirname(__file__), "..", "..", "job_scripts", "menu_scraper.log")
        with open(log_path, "a") as f:
            f.write(f"{now_ist().isoformat()} [API] {msg}\n")
    except Exception:
        pass

class MenuWeekChangeRequest(BaseModel):
    password: str
    number: int

@router.get("/")
async def get_mess_menu():
    try:    
        dir = os.path.dirname(os.path.realpath(__file__))
        with open(dir + "/config.json") as file:
            week = json.load(file)["week"]
        
        with open(dir + f"/{week}.json") as file:
            menu = json.load(file)
        
        # Check for special dinner override for today (IST)
        special_file = os.path.join(dir, "special_dinners.json")
        if os.path.exists(special_file):
            with open(special_file) as f:
                specials = json.load(f)
            today_str = today_ist().strftime("%d-%m-%Y")
            if today_str in specials:
                today_name = get_today_name()
                special_items = specials[today_str]
                # Override dinner for both messes
                if today_name in menu.get("LDH", {}):
                    menu["LDH"][today_name]["Dinner"] = special_items
                if today_name in menu.get("UDH", {}):
                    menu["UDH"][today_name]["Dinner"] = special_items
                log_api(f"Special dinner applied for {today_str} ({today_name}): {len(special_items)} items")
        
        log_api(f"Menu served for week={week}, date={today_ist()}")
        return menu
    except FileNotFoundError as e:
        log_api(f"Menu file not found: {e}")
        raise HTTPException(
            status_code=500, detail="Mess menu file does not exist. Please make one."
        )

@router.post("/")
async def post_mess_menu(admin: MenuWeekChangeRequest, request: Request):
    isAdmin = False
    user_id = None
    
    try:
        token = request.cookies.get("session")
        if token:
            status, data = verify_token(token)
            if status:
                user_id = data["sub"] 
        
        if user_id:
            user_details = get_user(user_id=user_id)
        if user_details['email'] in admins:
            isAdmin = True
    except HTTPException:
        isAdmin = False 
        
    if not isAdmin and admin.password != password:
        raise HTTPException(status_code=401, detail="Incorrect password")
    
    if admin.number not in allowed_numbers:
        raise HTTPException(status_code=400, detail="Invalid week number")
    
    dir = os.path.dirname(os.path.realpath(__file__))
    with open(dir + "/config.json", "w") as file:
        json.dump({"week": admin.number}, file, indent=1)
    
    log_api(f"Admin week override: week={admin.number} by user={user_id}")
    return {"message": "Week number updated successfully"}

@router.get("/week")
async def get_current_week_number(request: Request):
    isAdmin = False
    user_id = None
    
    try:
        user_id = get_user_id(request)
        user_details = get_user(user_id=user_id)
        if user_details['email'] in admins:
            isAdmin = True
    except HTTPException:
        isAdmin = False 
    
    if not isAdmin :
        return {"message" :"unauthorized"} 
    try:    
        dir = os.path.dirname(os.path.realpath(__file__))
        with open(dir + "/config.json") as file:
            week = json.load(file)["week"]
            
        return {"week": week}
    except FileNotFoundError:
        raise HTTPException(
            status_code=500, detail="Mess menu file does not exist. Please make one."
        )

@router.post("/webhook/sheets")
async def sheets_webhook(request: Request, background_tasks: BackgroundTasks):
    token = request.headers.get("X-Webhook-Token")
    if SHEETS_WEBHOOK_TOKEN and token != SHEETS_WEBHOOK_TOKEN:
        log_api(f"Webhook rejected: invalid token from {request.client.host if request.client else 'unknown'}")
        raise HTTPException(status_code=401, detail="Invalid webhook token")
    
    log_api(f"Webhook received from {request.client.host if request.client else 'unknown'}, triggering scraper")
    background_tasks.add_task(run_menu_scraper)
    return {"status": "scraper triggered"}

def run_menu_scraper():
    try:
        base_dir = os.path.dirname(os.path.realpath(__file__))
        scraper_path = os.path.join(base_dir, "..", "..", "job_scripts", "menu_scraper.py")
        result = subprocess.run(
            ["python", scraper_path],
            capture_output=True,
            text=True,
            timeout=120,
            cwd=os.path.join(base_dir, "..", "..")
        )
        if result.returncode != 0:
            log_api(f"Scraper FAILED: {result.stderr[:500]}")
        else:
            log_api(f"Scraper SUCCESS: {result.stdout[:200]}")
    except subprocess.TimeoutExpired:
        log_api("Scraper TIMEOUT after 120 seconds")
    except Exception as e:
        log_api(f"Scraper ERROR: {e}")